ROS howto guide (#51)
rerun-io/rerun-docs: docs/howto/ros2-nav-turtlebot.md

---

> title: Using Rerun with ROS2

Reviewer: ```suggestion
title: Using Rerun with ROS 2
```
The official way is to write "ROS 2"

order: 1
---

> Rerun does not have native ROS support, but many of the concepts between ROS and Rerun 
> line up fairly well. In this guide will show you how to write a simple ROS2 python node

Reviewer: ```suggestion
Rerun does not have native ROS support yet, but many of the concepts in ROS and Rerun 
line up fairly well. In this guide will show you how to write a simple ROS 2 python node
```
Perhaps also add a link to a Github issue about native ROS support in Rerun

that subscribes to some common ROS topics and logs them to Rerun.

> The following is primarily intended for existing ROS2 users. It will not spend much time
> covering how to use ROS2 itself.  If you are a Rerun user that is curious about ROS,
> please consult the [ROS2 Documentation](https://docs.ros.org/en/humble/index.html) instead.

Reviewer: ```suggestion
The following is primarily intended for existing ROS 2 users. It will not spend much time
covering how to use ROS 2 itself.  If you are a Rerun user that is curious about ROS,
please consult the [ROS 2 Documentation](https://docs.ros.org/en/humble/index.html) instead.
```

All of the code for this guide can be found on Github in

> [rerun/examples/python/ros](https://github.com/rerun-io/rerun/blob/main/examples/python/ros/).

Reviewer: Does it make sense to link to a specific commit?

Author: I'd generally rather have them get the latest code at risk of possibility of dead link. Added a comment here: https://github.com/rerun-io/rerun/issues/1345#issuecomment-1460738489

> ![Preview](/docs-media/ros1_preview.png)

Reviewer: I haven't written about this anywhere before but I think it's a good idea if we get into the habit of adding good alt texts. It's better for SEO and easy enough to do while adding new content.
```suggestion
![ROS 2 turtlebot3 gazebo simulation in a Rerun 3D view](/docs-media/ros1_preview.png)
```

### ROS2 & navigation
You will need to have installed [ROS2 Humble Hawksbill](https://docs.ros.org/en/humble/index.html)
and the [turtlebot3 navigation getting-started example](https://navigation.ros.org/getting_started/index.html).

Installing ROS is outside the scope of this guide, but you will need the equivalent of the following packages:
```
$ sudo apt install ros-humble-desktop gazebo ros-humble-navigation2 ros-humble-turtlebot3 ros-humble-turtlebot3-gazebo
```

Before proceeding, you should follow the [navigation example](https://navigation.ros.org/getting_started/index.html)
and confirm that you can successfully run:
```
$ export TURTLEBOT3_MODEL=waffle
$ export GAZEBO_MODEL_PATH=/opt/ros/humble/share/turtlebot3_gazebo/models
$ ros2 launch nav2_bringup tb3_simulation_launch.py headless:=False
```

Make sure that you can set the 2D Pose Estimate and send a Navigation Goal via rviz. You can now leave this
running in the background for the remainder of the guide.

### Additional Dependencies

The code for this guide is in the `rerun` repository. If you do not already have rerun cloned,
you should check it out now:
```
git clone git@github.com:rerun-io/rerun.git

> cd rerun

Reviewer: Checkout a specific commit?

Reviewer: Actually, maybe not

```
The example code can be found in the folder: `examples/python/ros`.

In addition to the ROS dependencies, the Rerun node makes use of some dependencies specified in
[`requirements.txt`](https://github.com/rerun-io/rerun/blob/main/examples/python/ros/requirements.txt).

Rerun recommends using `venv` (or the equivalent) to create an environment for installing these
dependencies.  Note that *after* setting up your virtualenv you will need to activate your ROS2
environment.
```
$ python3 -m venv venv
$ source venv/bin/active
(venv) $ pip install -r examples/python/ros/requirements.txt
(venv) $ source /opt/ros/humble/setup.bash
```

## Running the example

With the previous dependencies installed, and gazebo running, you should now be able to launch the Rerun ROS example:
```
(venv) $ python3 examples/python/ros/main.py
```

You should see a window similar to:

> ![Preview](/docs-media/ros2_launched.png)

Reviewer: Same here re alt text

### Overview

Assuming you are familiar with the turtlebot nav example and rviz, this view should be fairly familiar to you:

 * `map/box` is a placeholder for the map. (This will eventually be a map: [#1531](https://github.com/rerun-io/rerun/issues/1531)).
 * `map/robot` is a transform representing the robot pose logged as a
   [rigid3 transform](../reference/primitives#transform)
 * `map/robot/urdf` contains the `URDF` logged as a [mesh3d](../reference/primitives#mesh)

>  * `map/robot/scan` contains a `LaserScan` msg logged as a

Reviewer: Refer to an issue for adding a primitive to Rerun

    """Load a URDF file using `yourdfpy` and find resources via ament."""
    f = io.StringIO(msg.data)
    return URDF.load(f, filename_handler=ament_locate_package)
```

We then use `rerun_urdf.load_urdf_from_msg` from the URDF subscription callback.

Note that when developing this guide, we noticed that the camera mesh URDF was not having
its scale applied to it. This seems like a bug in either `yourdfpy` or `pycollada`
not respecting the scale hint.  To accomodate this, we manually re-scale the
camera link.

Once we have correctly re-scaled the camera component, we can send the whole scene to rerun with
`rerun_urdf.log_scene`.
```
def urdf_callback(self, urdf_msg: String) -> None:
    """Log a URDF using `log_scene` from `rerun_urdf`."""
    urdf = load_urdf_from_msg(urdf_msg)

    # The turtlebot URDF appears to have scale set incorrectly for the
    # camera_link. Although rviz loads it properly `yourdfpy` does not.
    orig, _ = urdf.scene.graph.get("camera_link")
    scale = trimesh.transformations.scale_matrix(0.00254)
    urdf.scene.graph.update(frame_to="camera_link", matrix=orig.dot(scale))
    scaled = urdf.scene.scaled(1.0)

>     log_scene(scene=scaled,
>               node=urdf.base_link,
>               path="map/robot/urdf",
>               timeless=True)

Reviewer: ```suggestion
    rerun_urdf.log_scene(scene=scaled,
                         node=urdf.base_link,
                         path="map/robot/urdf",
                         timeless=True)
```

```

Back in `rerun_urdf.log_scene` all the code is doing is recursively walking through
the trimesh scene graph and for each node, it extracting the transform to the parent,
which it logs via `rr.log_rigid3`, and then as the new path, using `rr.log_mesh` to
send the vertices, indices, and normals from the trimesh geometry:
```
node_data = scene.graph.get(frame_to=node, frame_from=parent)

if node_data:
    # Log the transform between this node and its direct parent (if it has one!).
    if parent:
        world_from_mesh = node_data[0]
        t = trimesh.transformations.translation_from_matrix(world_from_mesh)
        q = trimesh.transformations.quaternion_from_matrix(world_from_mesh)

>         # `trimesh` stores quaternions in `wxyz` format, rerun needs `xyzw`

Reviewer: Consider including the TODO here about not having to do the conversion once https://github.com/rerun-io/rerun/issues/883 is closed
